tests/Core/EngineTest.php

```diff
@@ -72,14 +72,41 @@ public function testCrawlsStartUrls(): void
             $this->makeRequest('http://localhost:8000/test1'),
             $this->makeRequest('http://localhost:8000/test2'),
         ];
-        $run = new Run($startRequests);
+        $run = new Run($startRequests, 'DefaultSpider');
 
         $this->engine->start($run);
 
         $this->assertRouteWasCrawledTimes('/test1', 1);
         $this->assertRouteWasCrawledTimes('/test2', 1);
     }
 
+    public function testDoesntCrawlStartUrlsWithExistingRequestsInScheduler(): void
+    {
+        $startRequests = [
+            $this->makeRequest('http://localhost:8000/test1'),
+            $this->makeRequest('http://localhost:8000/test2'),
+        ];
+
+        $run = new Run($startRequests, 'DefaultSpider');
+
+        $dispatcher = new EventDispatcher();
+        $scheduler = new ArrayRequestScheduler($this->clock);
+        $scheduler->schedule($this->makeRequest('http://localhost:8000/test3'));
+        $engine = new Engine(
+            $scheduler,
+            new Downloader(new Client(), $dispatcher),
+            new ItemPipeline($dispatcher),
+            new Processor($dispatcher),
+            $dispatcher,
+        );
+
+        $engine->start($run);
+
+        $this->assertRouteWasCrawledTimes('/test1', 0);
+        $this->assertRouteWasCrawledTimes('/test2', 0);
```

Comment:
These assertions are actually failing because of a bug in the `assertRouteWasCrawledTimes` method. The easiest fix would be to use `assertRouteWasNotCrawled` instead.

```suggestion
        $this->assertRouteWasNotCalled('/test1');
        $this->assertRouteWasNotCalled('/test2');
```

Reply:
sorry i wrote this while very tired, i'll look over things before committing again